Trim complexity in CommentFactoryBase

janekdb opened the pull request:
1. Replace option handling with library call
2. Remove NumberFormatException catch presumed to be copy/paste legacy
3. It's Scaladoc
4. Parse trailing whitespace regex once instead of per line
5. Use string interpolation where it improves readability
6. Add missed alternative to block grammar production rule
7. Add regression test for tag requirement warnings

janekdb commented:
I need some advice on how to add a test for the modified warning code. The message should be unchanged. This object covers one path,

```
13:37 $ cat DoubleVersionTag.scala

 /**
  * @version 1.0
  * @version 2.0
  */
object DoubleVersionTag
✔ ~/workspaces/public/scala [2.11.x-CommentFactoryBase-mix-1 ↑·1|…102⚑ 18] 
13:37 $ ./build/quick/bin/scaladoc DoubleVersionTag.scala 
model contains 2 documentable templates
DoubleVersionTag.scala:2: warning: Only one '@version' tag is allowed
 /**
 ^
one warning found
```

janekdb commented:
Review by @VladUreche, @gourlaysama.

janekdb commented:
This is the build failure,

```
[WARNING] The requested profile "scala-2.11.x" could not be activated because it does not exist.
[ERROR] The goal you specified requires a project to execute but there is no POM in this directory (/home/jenkins/workspace/scala-2.11.x-integrate-ide/target/scalariform). Please verify you invoked Maven from the correct directory. -> [Help 1]
org.apache.maven.lifecycle.MissingProjectException: The goal you specified requires a project to execute but there is no POM in this directory (/home/jenkins/workspace/scala-2.11.x-integrate-ide/target/scalariform). Please verify you invoked Maven from the correct directory.
```

janekdb commented:
/rebuild

VladUreche commented:
@janekdb Thank you for the cleanup, it's always appreciated! LGTM

Regarding your questions, I'm a bit surprised you don't get the warning... There [a redirect in `ScaladocModelTest`](https://github.com/scala/scala/blob/2.12.x/src/scaladoc/scala/tools/partest/ScaladocModelTest.scala#L62-L64) which should guarantee that anything printed ends up captured by the test.

Unfortunately atm I just don't have the bandwidth to debug this. Would you mind filing a bug and assigning it to me?

janekdb commented:
@VladUreche Thanks for the feedback. It transpires that extracting the comments triggers the expected warning.

```
  def testModel(root: Package) = {
    // get the quick access implicit defs in scope (_package(s), _class(es), _trait(s), object(s) _method(s), _value(s))
    import access._

    val base = root._package("scala")._package("test")._package("scaladoc")._package("tagrequirements")

    val TestObject = base._object("TestObject")
    val testObjectComment = extractCommentText(TestObject.comment.get)
    println("hi")
    println(testObjectComment)
    println("bye")
  }
```

Next step is for me to minimise the above code and then we should be good to add the test.

VladUreche commented:
@janekdb oh, yeah, the text is only parsed when the doc comment is requested. How stupid of me not to think about that... Thanks for debugging that!

VladUreche commented:
It should even be enough to say `TestObject.comment`, that is enough to trigger parsing.

janekdb commented on test/scaladoc/run/tag-requirements.check:
```diff
@@ -0,0 +1,16 @@
+newSource:9: warning: Tag '@param' must be followed by a symbol name
+          /**
+          ^
+newSource:9: warning: Tag '@param' is not recognised
```
I think we could raise a later improvement which removes this warning on the basis that the `param` tag has already been warned on. It's not actually unrecognised but unused because of the missing symbol name.

VladUreche replied:
Agreed.

VladUreche commented:
/rebuild

SethTisue commented:
/nothingtoseehere

(integrate-ide failure is unrelated to the PR)

lrytz commented on src/scaladoc/scala/tools/nsc/doc/base/CommentFactoryBase.scala:
```diff
@@ -334,15 +332,16 @@ trait CommentFactoryBase { this: MemberLookupBase =>
           val pairs: Seq[(String, Body)] =
             for (key <- keys) yield {
               val bs = (bodyTags remove key).get
+              // > 0 ?? - JDB
```
@janekdb `> 1` seems correct to me

janekdb replied:
I'll remove this comment. Code is correct as-is.

janekdb committed:
Trim complexity in CommentFactoryBase

1. Replace option handling with library call
2. Remove NumberFormatException catch presumed to be copy/paste legacy
3. It's Scaladoc
4. Parse trailing whitespace regex once instead of per line
5. Use string interpolation where it improves readability
6. Add missed alternative to block grammar production rule
7. Add regression test for tag requirement warnings

SethTisue commented:
/rebuild

SethTisue commented:
wow, it is not easy to figure out where the actual failure is in 18MB of log. @sschaef, can you tell what's going on here?

sschaef commented:
/rebuild

sschaef commented:
Looks like a flaky test - at least I can't see how this PR should affect such a test. I rerun the tests once again. Maybe this time they pass.

SethTisue commented:
(I see Jenkins is backed up right now, looks like we might need to wait a few hours before this will re-run.)

lrytz commented:
LGTM, thanks @janekdb!
